src/librbd/io/SimpleSchedulerObjectDispatch.h

```diff
@@ -0,0 +1,148 @@
+      uint64_t* journal_tid, io::DispatchResult* dispatch_result,
+      Context** on_finish, Context* on_dispatched) override;
+
+  bool write(
+      const std::string &oid, uint64_t object_no, uint64_t object_off,
+      ceph::bufferlist&& data, const ::SnapContext &snapc, int op_flags,
+      const ZTracer::Trace &parent_trace, int* object_dispatch_flags,
+      uint64_t* journal_tid, io::DispatchResult* dispatch_result,
+      Context** on_finish, Context* on_dispatched) override;
+
+  bool write_same(
+      const std::string &oid, uint64_t object_no, uint64_t object_off,
+      uint64_t object_len, io::Extents&& buffer_extents,
+      ceph::bufferlist&& data, const ::SnapContext &snapc, int op_flags,
+      const ZTracer::Trace &parent_trace, int* object_dispatch_flags,
+      uint64_t* journal_tid, io::DispatchResult* dispatch_result,
+      Context** on_finish, Context* on_dispatched) override;
+
+  bool compare_and_write(
+      const std::string &oid, uint64_t object_no, uint64_t object_off,
+      ceph::bufferlist&& cmp_data, ceph::bufferlist&& write_data,
+      const ::SnapContext &snapc, int op_flags,
+      const ZTracer::Trace &parent_trace, uint64_t* mismatch_offset,
+      int* object_dispatch_flags, uint64_t* journal_tid,
+      io::DispatchResult* dispatch_result, Context** on_finish,
+      Context* on_dispatched) override;
+
+  bool flush(
+      io::FlushSource flush_source, const ZTracer::Trace &parent_trace,
+      io::DispatchResult* dispatch_result, Context** on_finish,
+      Context* on_dispatched) override;
+
+  bool invalidate_cache(Context* on_finish) override {
+    return false;
+  }
+  bool reset_existence_cache(Context* on_finish) override {
+    return false;
+  }
+
+  void extent_overwritten(
+      uint64_t object_no, uint64_t object_off, uint64_t object_len,
+      uint64_t journal_tid, uint64_t new_journal_tid) override {
+  }
+
+private:
+  struct ObjectDelayedRequests {
+    uint64_t object_no;
+    utime_t dispatch_time;
+    uint64_t offset = 0;
+    ceph::bufferlist data;
+    SnapContext snapc = {0, {}};
+    int op_flags = 0;
+    int object_dispatch_flags = 0;
+    std::list<Context *> requests;
+    uint64_t dispatch_seq = 0;
+
+    ObjectDelayedRequests(uint64_t object_no) : object_no(object_no) {
+    }
+  };
+
+  typedef std::map<uint64_t, ObjectDelayedRequests> DelayedRequests;
+
+  ImageCtxT* m_image_ctx;
+
+  Mutex m_lock;
+  SafeTimer *m_timer;
+  Mutex *m_timer_lock;
+  uint64_t m_max_delay;
+  uint64_t m_dispatch_seq = 0;
+
+  DelayedRequests m_delayed_requests;
+  std::list<ObjectDelayedRequests *> m_dispatch_queue;
+  Context *m_timer_task = nullptr;
+
+  bool can_merge_write(uint64_t object_no, uint64_t object_off,
+                       ceph::bufferlist&& data, const ::SnapContext &snapc,
+                       int op_flags, int object_dispatch_flags,
+                       Context* on_dispatched);
+
+  void dispatch_delayed_requests(Context** on_finish);
```

Comment:
Nit: at a glance, it's pretty confusing to have all so many different permutations of different parameters to the same method name. Perhaps this one could be ```dispatch_all_delayed_requests```>